### PR TITLE
docs: Add documentation about memory in Cloud-Hypervisor

### LINKED ISSUE #1659: Write documentation for Cloud-Hypervisor memory and NUMA
We recently merged multiple patches refactoring the memory manager and introducing the NUMA support, therefore it'd be important to write some documentation on how to use all these options.

### PR DESCRIPTION
**Pull request by sboeuf**
Writing some new documentation to help users understand how the guest
memory can be described through Cloud-Hypervisor parameters.

Fixes #1659

Signed-off-by: Sebastien Boeuf <sebastien.boeuf@intel.com>

### CODE REVIEW

File: docs/memory.md [hunk 1]
```diff
@@ -0,0 +1,389 @@
+# Memory
+
+Cloud-Hypervisor has many ways to expose memory to the guest VM. This document
+aims at explaining what Cloud-Hypervisor is capable of and how it can be used
```
Comment (rbradford) [inline docs/memory.md:4]: s/aims at explaining/aims to explain/

File: docs/memory.md [hunk 2]
```diff
@@ -0,0 +1,389 @@
...
+to meet the needs from very different use cases.
```
Comment (rbradford) [inline docs/memory.md:5]: s/from/of/

File: docs/memory.md [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+--memory <memory>	Memory parameters "size=<guest_memory_size>,mergeable=on|off,shared=on|off,hugepages=on|off,hotplug_method=acpi|virtio-mem,hotplug_size=<hotpluggable_memory_size>,balloon=on|off"
+```
+
+### `size`
+
+Size of the RAM in the guest VM.
+
+This option is mandatory when using the `--memory` parameter.
+
+Value is an unsigned integer on 64 bits.
```
Comment (rbradford) [inline docs/memory.md:36]: s/on/of/

File: docs/memory.md [hunk 4]
```diff
@@ -0,0 +1,389 @@
...
+```
+
+### `mergeable`
+
+Specifies if the pages from the guest RAM must be marked as _mergeable_. In
+case this option is `true` or `on`, the pages will be marked with `madvise(2)`
+to let the host kernel know which pages are eligible for being merged by the
+KSM daemon.
+
+This option can be used when trying to reach the higher density of VMs running
```
Comment (rbradford) [inline docs/memory.md:51]: s/the higher/a higher/

File: docs/memory.md [hunk 5]
```diff
@@ -0,0 +1,389 @@
...
+### `shared`
+
+Specifies if the memory must be `mmap(2)` with `MAP_SHARED` flag.
+
+By sharing a memory mapping, one can share the guest RAM with other processes
+running on the host. One can use this option when running vhost-user devices
+as part of the VM device model, as they will be driven by standalone daemons
+needing access to the guest RAM content.
+
+By default this option is turned off, which result in performing `mmap(2)`
```
Comment (rbradford) [inline docs/memory.md:71]: s/which result/which results/

File: docs/memory.md [hunk 6]
```diff
@@ -0,0 +1,389 @@
...
+memory from the host. Assuming a host has two types of memory, with one slower
+than the other, each related to a distinct NUMA node, one could create a VM
+with slower memory accesses by backing the entire guest RAM from the furthest
+NUMA node on the host.
+
+This option also gives the opportunity to create a VM with non uniform memory
+accesses as one could define a first memory zone backed by fast memory, and a
+second memory zone backed by slow memory.
+
+Value is an unsigned integer on 32 bits.
```
Comment (rbradford) [inline docs/memory.md:263]: s/on/of/

File: docs/memory.md [hunk 7]
```diff
@@ -0,0 +1,389 @@
...
+Node identifier of a node that must be created in the guest. This option gives
+the user a way to create NUMA nodes in the guest and associate them with memory
+zones.
+
+This option can be very useful and powerful when combined with `host_numa_node`
+as it allows for creating a VM with non uniform memory accesses, and let the
+guest know about it. It allows for exposing memory zones through different NUMA
+nodes, which can help the guest workload run more efficiently.
+
+Value is an unsigned integer on 32 bits.
```
Comment (rbradford) [inline docs/memory.md:283]: Difto

File: docs/memory.md [hunk 8]
```diff
@@ -0,0 +1,389 @@
...
+
+### `id`
+
+Node identifier of a guest NUMA node. The node referred by this identifier has
+been created through the `guest_numa_node` option from the `--memory-zone`
+parameter.
+
+This option is mandatory when using the `--numa` parameter.
+
+Value is an unsigned integer on 32 bits.
```
Comment (rbradford) [inline docs/memory.md:318]: Ditto

**Review by rbradford — APPROVED** [review]
👍.minor nits

**Comment by sboeuf** [comment]
Thanks for the review @rbradford, comments have been addressed.

**Pull request merged**